### PR TITLE
[ot_certs] ASN1 codegen

### PR DESCRIPTION
**Pull request by pamaury**
Only review the last three commits, others are from #20601. See commits for description.
To test:
```bash
# build all files
./bazelisk.sh build //sw/device/silicon_creator/lib/cert/...
# view all targets:
./bazelisk.sh query //sw/device/silicon_creator/lib/cert/...
# query list of output for a certificate:
./bazelisk.sh outquery-all //sw/device/silicon_creator/lib/cert:example_template
# use your favorite editor to look at those files
```

### CODE REVIEW

File: sw/host/ot_certs/src/codegen.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/codegen.rs:70]: Nit / suggestion: you may find it clearer to put this in a `.c` file in this directory and use `include_str!` to pull it in.

File: rules/certificates.bzl
Comment (jwnrt) [inline rules/certificates.bzl:41]: Would you mind adding these doc strings before we merge?

It could easily be forgotten if we leave it to the future

**Review by jwnrt — APPROVED** [review]
I've done a pass on this and it looks generally okay, but there's really lots of generation code and I definitely haven't checked its correctness in depth.

I don't have any refactorings to suggest, but if you have anything that could make this easier to follow or to make it more compact, that could be helpful for the future.

**Comment by pamaury** [comment]
Thanks for the comments James, I will rework that. Regarding the code generation, I agree that it's a lot of code but I don't have any great idea to make the code more compact. I'll think about it when updating the code.

**Comment by pamaury** [comment]
I have removed the code of the writer from this PR since it takes quite a bit of code and it is not essential. I have addressed some changes, I still need to fix two TODO.

File: sw/device/silicon_creator/lib/cert/BUILD
Comment (timothytrippel) [inline sw/device/silicon_creator/lib/cert/BUILD:12]:
```suggestion
    name = "generic_template",
    template = "generic.hjson",
```

applies throughout

File: rules/certificates.bzl
Comment (timothytrippel) [inline rules/certificates.bzl:44]:
```suggestion
# done by this rule. The generated files are put in the following
```

File: rules/certificates.bzl
Comment (timothytrippel) [inline rules/certificates.bzl:30]: we probably also want to hit the generated C files with clang format after they are generated?

**Review by timothytrippel — APPROVED** [review]

File: sw/host/ot_certs/src/asn1/codegen.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/asn1/codegen.rs:419]: Nit: can we put `(up to three bytes)` in parens?

File: sw/host/ot_certs/src/asn1/codegen.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/asn1/codegen.rs:421]: Nit: if the `1` is for the null terminator, can we put it at the end to make that clear?

File: sw/host/ot_certs/src/codegen.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/codegen.rs:41]: Nit: de-indent this one level

**Review by jwnrt — APPROVED** [review]
Reviewed this with @luismarques and we only had these nits

File: rules/certificates.bzl
Comment (pamaury) [reply rules/certificates.bzl:30]: I will look into it.

**Comment by pamaury** [comment]
I have pushed an update that rebases the code and also addresses the various comments. I have also improved the documentation and the code now produces better estimates for the size of the certificate (previous estimates were in the ball park of 550 bytes, now it is at 440 so it's much better!).

File: sw/host/ot_certs/src/codegen.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/codegen.rs:70]:
```suggestion
    write!(source_h, "#ifndef __{preproc_guard_include}__\n");
```

**Pull request merged**
